### PR TITLE
[#1548] - redefines PrincipalCollection interface to be intended as immutable

### LINKED ISSUE #1548: [Enhancement] Mutability of PrincipalCollection and AuthenticationInfo
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/shiro/issues?q=is%3Aissue) and found no similar issues.

### Enhancement Request

First of all, this is neither a feature request, nor is it strictly a bug report -- but the latter cannot be decided, as I will explain below. I am currently evaluating Shiro for a project and have found an architectural decision in its code that I believe is dangerous.

The culprit is PrincipalCollection. This interface does not itself specify whether it allows mutable implementations, but the existence of MutablePrincipalCollection does. Code that receives a PrincipalCollection from an unknown source must therefore treat it as "potentially mutable". PCs are passed around a lot, so "unknown source" is a common case. 

Unfortunately, receiving code doesn't do that. Received PCs are passed and stored as if they cannot be modified anymore. AbstractAuthenticator even stores them in a cache. All these places are affected by later mutations to the PC.

The code that does mutate a PC, AbstractAuthenticationStrategy (from the PAM package), will either accept an MPC from its caller and mutate that, or create its own copy in case the one from the caller is not mutable. This further obscures _which_ PC gets mutated.

The current code from Shiro does not show faulty behavior from this problem because AbstractAuthenticationStrategy is the only place where PCs get mutated, and all subclasses of that class except for FirstSuccessfulStrategy only mutate an MPC they created themselves (the "aggregate"). FirstSuccessfulStrategy uses a PC returned by a Realm as the aggregate, but then doesn't actually mutate it.

**Why is this a problem?**

I think this is a problem because it is bug-free by chance, not by design. It depends on the implementations of the interfaces used to make stronger guarantees than the interfaces do -- and these implementations don't even do that, only the combined behavior of the implementations of multiple interfaces do. 

While projects using Shiro could easily trip over this, my main concern is my own project being affected after upgrading to future Shiro versions, with its internals changing in such a way that it breaks one of those guarantees that the interfaces don't make but on which correctness of the code relies, and has to rely, on.

### Describe the solution you'd like

Possible solutions:

* My preferred solution: Make PrincipalCollection immutable by design. AFAIK, the PAM package doesn't really need it to be mutable and could instead make mutated copies. The biggest problem is code using Shiro that might make use of MPCs -- it's a breaking change.
* Make defensive copies of PCs everywhere. This is ugly, it's easy to forget a place, and it might start to affect performance.
* Establish rules about PC mutability. This is the non-invasive approach and might be a first step towards making PCs immutable.

An example for a "rule about mutability" could be: PCs should be treated as immutable, unless you created the MPC yourself and keep it to yourself -- or make a clear contract with your callers that the PC is treated as mutable. AbstractAuthenticationStrategy would have to be changed to adhere to that rule.

I am happy to contribute to the solution, but I first want to make sure that this is seen as a problem by others too, and discuss which solution approach is preferred.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by janitza-mage**
Closes #1548. Declare PrincipalCollection interface to be intended to be immutable; provide an immutable implementation including a builder class; deprecate the mutable implementation; use the immutable one in SimpleAuthenticationInfo for merging.

Background: Merging AuthenticationInfo and the contained PrincipalCollections previously lead to one of the involved PrincipalCollections be selected by undefined means and being mutated, propagating the changes to other callers that did not expect such changes, including the authentication cache.

Following this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [GitHub issue](https://github.com/apache/shiro/issues) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a GitHub issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[#XXX] - Fixes bug in SessionManager`,
       where you replace `#XXX` with the appropriate GitHub issue. Best practice
       is to use the GitHub issue title in the pull request title and in the first line of the commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] add `fixes #XXX` if merging the PR should close a related issue.
 - [x] Run `mvn verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If you have a group of commits related to the same change, please squash your commits into one and force push your branch using `git rebase -i`. 
 - [x] Committers: Make sure a milestone is set on the PR
 
Trivial changes like typos do not require a GitHub issue (javadoc, comments...). 
In this case, just format the pull request title like `[DOC] - Add javadoc in SessionManager`.
 
If this is your first contribution, you have to read the [Contribution Guidelines](https://github.com/apache/shiro/blob/master/CONTRIBUTING.md)

If your pull request is about ~20 lines of code you don't need to sign an [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) 
if you are unsure please ask on the developers list.

To make clear that you license your contribution under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Review by bdemers — CHANGES_REQUESTED** [review]
I took a quick pass through, it looks great! thank you!
One nit, can you add the `@Deprecated` annotation as well (to `MutablePrincipalCollection` anywhere it's used (and set `forRemoval=true`)
(and  possibly the `@deprecated` javadoc tag to `SimplePrincipalCollection`, with a note to use `ImmutablePrincipalCollection` instead.

Thanks again!!
(I'll trigger the CI builds for this)

**Comment by janitza-mage** [comment]
I have added @deprecated to SimplePrincipalCollection. That was the only place where MutablePrincipalCollection is referenced.

**Review by lprimak — APPROVED** [review]

**Comment by lprimak** [comment]
@janitza-mage Can you please restore the template for this PR?
We would need you to click appropriate checkboxes for a contribution license acceptance.

**Comment by janitza-mage** [comment]
I edited the initial comment. Remarks:
* The "fixes #XXX" is N/A because no related issue gets fixed by this
* about the milestone: I have no clue what is meant by that
* the code is apache-licensed, so I ticked the first box and left the second one unticked -- is that correct?

**Comment by lprimak** [comment]
Thanks.
I fixed up the remaining issues.

**Comment by janitza-mage** [comment]
Is there anything left for me to do in this PR, then? I think I can't merge it myself, right?

**Comment by lprimak** [comment]
No,you are good.
We need to create 3.x development branch, make some new infra etc.
It's gonna be a while until we merge this unfortunately.

**Pull request merged**
